third_party/move/move-compiler-v2/src/env_pipeline/cmp_rewriter.rs

```diff
@@ -0,0 +1,149 @@
+        let arg_ty = self.env.get_node_type(arg.as_ref().node_id());
+        let new_ref_type = Type::Reference(ReferenceKind::Immutable, Box::new(arg_ty.clone()));
+        let new_ref_id = self.env.new_node(arg_loc, new_ref_type);
+        ExpData::Call(
+            new_ref_id,
+            Operation::Borrow(ReferenceKind::Immutable),
+            vec![arg.clone()],
+        )
+        .into_exp()
+    }
+
+    // We only transform struct/vector arguments
+    // Other types are omitted as copying them can be cheaper than creating a reference
+    fn arg_can_transform(&mut self, arg: &Exp) -> bool {
+        let arg_ty = self.env.get_node_type(arg.as_ref().node_id());
+        matches!(arg_ty, Type::Vector(_)) || matches!(arg_ty, Type::Struct(_, _, _))
+    }
+
+    fn ref_from_derefed_arg(&mut self, arg: &Exp) -> Option<Exp> {
+        if let ExpData::Call(_, Operation::Deref, deref_args) = arg.as_ref() {
+            debug_assert!(
+                deref_args.len() == 1,
+                "there should be exactly one argument for dereference"
+            );
+            match deref_args[0].as_ref() {
+                // Deref(Borrow(Immutable))
+                ExpData::Call(_, Operation::Borrow(ReferenceKind::Immutable), ..) => {
+                    return Some(deref_args[0].clone());
+                },
+                // Deref(vector::borrow)
+                ExpData::Call(_, Operation::MoveFunction(mid, fid), ..) => {
+                    // If the deref argument is a MoveFunction call, we check if it is a borrow
+                    if self.env.get_module(*mid).matches_name("vector")
+                        && self
+                            .env
+                            .get_function(mid.qualified(*fid))
+                            .matches_name("borrow")
+                    {
+                        return Some(deref_args[0].clone());
+                    }
+                },
+                // Todo: we should also optimize Deref(global_borrow)
+                _ => {
+                    return None;
+                },
+            }
+        }
+        None
+    }
+}
+
+impl ExpRewriterFunctions for CmpRewriter<'_> {
+    fn rewrite_call(&mut self, call_id: NodeId, oper: &Operation, args: &[Exp]) -> Option<Exp> {
+        match oper {
+            Operation::Eq | Operation::Neq => {
+                // Check if all arguments are tranformable (i.e., are values)
+                let transformable: Vec<bool> =
+                    args.iter().map(|exp| self.arg_can_transform(exp)).collect();
+                if transformable.iter().all(|x| *x) {
+                    let transformed_args =
+                        args.iter().map(|exp| self.rewrite_cmp_arg(exp)).collect();
+                    // Create a new call expression with the transformed arguments
+                    return Some(ExpData::Call(call_id, oper.clone(), transformed_args).into_exp());
+                }
+                None
```

Comment:
nit: use `else None` and skip the `return` to get a more readable (functional) style.